staging/src/k8s.io/apiserver/pkg/server/options/audit.go

```diff
@@ -525,6 +526,9 @@ func (o *AuditLogOptions) getWriter() (io.Writer, error) {
 }
 
 func (o *AuditLogOptions) ensureLogFile() error {
+	if err := os.MkdirAll(filepath.Dir(o.Path), 0700); err != nil {
```

Comment:
I'm unsure whether we want to do this, and if these are the right permissions... if kube-apiserver is usually run in a container, and the directory where the log file is placed is expected to be visible outside the container, will creating this in-process cause problems?

Reply:
good point, i agree with Jordan.

Reply:
The log file is already with `0600` permissions, and I the `x` permissions are required if more than 1 directory needs to be created. Here is a short example that none of `0400`, `0500`, `0600` provides sufficient permissions
```golang
package main

import (
	"fmt"
	"io/fs"
	"io/ioutil"
	"os"
	"path/filepath"
)

func testFileAndDirCreation(path string, dirMode fs.FileMode) error {
	dirName := filepath.Dir(path)
	if err := os.MkdirAll(dirName, dirMode); err != nil {
		return err
	}
	return ioutil.WriteFile(path, []byte("test"), dirMode)
}

func main() {
	dirModes := []fs.FileMode{0400, 0500, 0600, 0700}
	for _, dirMode := range dirModes {
		tmpPath, err := os.MkdirTemp("", "permission-test")
		if err != nil {
			fmt.Printf("Failed to create tmp dir\n")
			os.Exit(1)
		}
		path := filepath.Join(tmpPath, os.Args[1])
		err = testFileAndDirCreation(path, dirMode)
		if err != nil {
			fmt.Printf("Failed to create file %s with dirMode %o, err: %v\n", path, dirMode, err)
		} else {
			fmt.Printf("Successfully created file %s with dirMode %o, err: %v\n", path, dirMode, err)
		}
	}
}
```

A sample output is
```bash
$ go run main.go /test/dirmode/without/x/permissions/audit.log
Failed to create file /tmp/permission-test4032771709/test/dirmode/without/x/permissions/audit.log with dirMode 400, err: mkdir /tmp/permission-test4032771709/test/dirmode: permission denied
Failed to create file /tmp/permission-test535255279/test/dirmode/without/x/permissions/audit.log with dirMode 500, err: mkdir /tmp/permission-test535255279/test/dirmode: permission denied
Failed to create file /tmp/permission-test4510376/test/dirmode/without/x/permissions/audit.log with dirMode 600, err: mkdir /tmp/permission-test4510376/test/dirmode: permission denied
Successfully created file /tmp/permission-test2117089606/test/dirmode/without/x/permissions/audit.log with dirMode 700, err: <nil>

```

> if kube-apiserver is usually run in a container, and the directory where the log file is placed is expected to be visible outside the container, will creating this in-process cause problems?

In such scenarios the directories can be created in advance with the required permissions and `os.MkdirAll` will not do anything.

Reply:
@dims @liggitt can you please elaborate?

With k8s 1.21 I am able to specify path with non-existing directories for the `--audit-log-path` flag and k8s (actually the lumberjack module) takes care to create the directories. The permissions for the dirs and files in the path `/var/lib/audit1/audit2/audit3/audit.log`, where `audit1`, `audit2` and `audit3` are not existing at the time `kube-apiserver` is started, are:
```bash
/var/lib# ls -lRa
.:
total 16
drwxr-xr-x 3 root root 4096 Jun 28 07:39 .
drwxr-xr-x 3 root root 4096 Jan  1  1970 ..
drwxr--r-- 3 root root 4096 Jun 28 07:39 audit1

./audit1:
total 12
drwxr--r-- 3 root root 4096 Jun 28 07:39 .
drwxr-xr-x 3 root root 4096 Jun 28 07:39 ..
drwxr--r-- 3 root root 4096 Jun 28 07:39 audit2

./audit1/audit2:
total 12
drwxr--r-- 3 root root 4096 Jun 28 07:39 .
drwxr--r-- 3 root root 4096 Jun 28 07:39 ..
drwxr--r-- 2 root root 4096 Jun 28 07:39 audit3

./audit1/audit2/audit3:
total 70796
drwxr--r-- 2 root root     4096 Jun 28 07:39 .
drwxr--r-- 3 root root     4096 Jun 28 07:39 ..
-rw-r--r-- 1 root root 72480850 Jun 28 07:51 audit.log
```

If I try the same with k8s 1.22, `kube-apiserver fails with 
```
Error: ensureLogFile: open /var/lib/audit1/audit2/audit3/audit.log: no such file or directory
```

And with the change in this PR the permissions will be `0700` which is still restricted compared to previous permissions `0755`. 

From this point of view, I am actually fixing an regression that was introduced with #95387.

Reply:
I've built an image with the change from this PR, started the otherwise failing `kube-apiserver` and inspected the created directories and files, so the result permissions are exactly as expected `0700`:

```bash
/var/lib# ls -lRa
.:
total 24
drwxr-xr-x 1 root root 4096 Jul  1 16:23 .
drwxr-xr-x 1 root root 4096 Jan  1  1970 ..
drwx------ 3 root root 4096 Jul  1 16:23 audit1

./audit1:
total 12
drwx------ 3 root root 4096 Jul  1 16:23 .
drwxr-xr-x 1 root root 4096 Jul  1 16:23 ..
drwx------ 3 root root 4096 Jul  1 16:23 audit2

./audit1/audit2:
total 12
drwx------ 3 root root 4096 Jul  1 16:23 .
drwx------ 3 root root 4096 Jul  1 16:23 ..
drwx------ 2 root root 4096 Jul  1 16:23 audit3

./audit1/audit2/audit3:
total 8
drwx------ 2 root root 4096 Jul  1 16:23 .
drwx------ 3 root root 4096 Jul  1 16:23 ..
-rw------- 1 root root    0 Jul  1 16:23 audit.log
```

Reply:
thanks for checking. for completeness, can you run the same test on the commit just before https://github.com/kubernetes/kubernetes/pull/95387 merged? if it's the same, then this looks reasonable to me